cmd/util/cmd/epochs/cmd/recover.go

```diff
@@ -0,0 +1,186 @@
+	generateRecoverEpochTxArgsCmd.Flags().StringVar(&flagPartnerWeights, "partner-weights", "", "path to a JSON file containing "+
+		"a map from partner node's NodeID to their stake")
+	generateRecoverEpochTxArgsCmd.Flags().Uint64Var(&flagStartView, "start-view", 0, "start view of the recovery epoch")
+	generateRecoverEpochTxArgsCmd.Flags().Uint64Var(&flagStakingEndView, "staking-end-view", 0, "end view of the staking phase of the recovery epoch")
+	generateRecoverEpochTxArgsCmd.Flags().Uint64Var(&flagEndView, "end-view", 0, "end view of the recovery epoch")
+}
+
+// generateRecoverEpochTxArgs generates recover epoch transaction arguments from a root protocol state snapshot and writes it to a JSON file
+func generateRecoverEpochTxArgs(cmd *cobra.Command, args []string) {
+	stdout := cmd.OutOrStdout()
+
+	// get flow client with secure client connection to download protocol snapshot from access node
+	config, err := common.NewFlowClientConfig(flagAnAddress, flagAnPubkey, flow.ZeroID, false)
+	if err != nil {
+		log.Fatal().Err(err).Msg("failed to create flow client config")
+	}
+
+	flowClient, err := common.FlowClient(config)
+	if err != nil {
+		log.Fatal().Err(err).Msg("failed to create flow client")
+	}
+
+	snapshot, err := common.GetSnapshot(context.Background(), flowClient)
+	if err != nil {
+		log.Fatal().Err(err).Msg("failed")
+	}
+
+	// extract arguments from recover epoch tx from snapshot
+	txArgs := extractRecoverEpochArgs(snapshot)
+
+	// encode to JSON
+	encodedTxArgs, err := epochcmdutil.EncodeArgs(txArgs)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not encode recover epoch transaction arguments")
+	}
+
+	// write JSON args to stdout
+	_, err = stdout.Write(encodedTxArgs)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not write jsoncdc encoded arguments")
+	}
+}
+
+// extractResetEpochArgs extracts the required transaction arguments for the `resetEpoch` transaction
+func extractRecoverEpochArgs(snapshot *inmem.Snapshot) []cadence.Value {
+	epoch := snapshot.Epochs().Current()
+	ids, err := epoch.InitialIdentities()
+	if err != nil {
+		log.Fatal().Err(err).Msg("failed to get initial identities for current epoch")
+	}
+
+	currentEpochDKG, err := epoch.DKG()
+	if err != nil {
+		log.Fatal().Err(err).Msg("failed to get DKG for current epoch")
+	}
+
+	log.Info().Msg("collecting partner network and staking keys")
+	partnerNodes := common.ReadPartnerNodeInfos(log, flagPartnerWeights, flagPartnerNodeInfoDir)
+	log.Info().Msg("")
+
+	log.Info().Msg("generating internal private networking and staking keys")
+	internalNodes := common.ReadInternalNodeInfos(log, flagInternalNodePrivInfoDir, flagConfig)
+	log.Info().Msg("")
+
+	log.Info().Msg("computing collection node clusters")
+	_, clusters, err := common.ConstructClusterAssignment(log, partnerNodes, internalNodes, flagCollectionClusters)
+	if err != nil {
+		log.Fatal().Err(err).Msg("unable to generate cluster assignment")
+	}
```

Comment:
> - Exit with an error if we observe a node in `InternalNodeInfos` that doesn't exist in the snapshot
> - Populate `internalNodes` based on what we read from disk
> - Populate `partnerNodes` based on `snapshot.Identities - internalNodes`

Very much possible that we are suggesting the same thing, but I am not sure. I generally agree with [Jordan's comment](https://github.com/onflow/flow-go/pull/5576#discussion_r1536041068), but I would frame it slightly differently: 
* The `snapshot` is the core object for this logic here:
   * It specifies the set of nodes $S$ which nodes are currently allowed to participate in the network. Note that ejected nodes are not part of $S$. So you **cannot  use `InitialIdentities`** as you do here https://github.com/onflow/flow-go/blob/7d09f47d73fbf7930a38d2e591d783209c7f92a2/cmd/util/cmd/epochs/cmd/recover.go#L105 Instead, you should use the `Identities` method: https://github.com/onflow/flow-go/blob/d89c1c30a0384a15d96272cd47d4e724b0639372/state/protocol/snapshot.go#L56
   * If you use a selector that drops ejected nodes and nodes with zero weight, you receive the set $S$ of all participating nodes as of this block.    
* The nodes in the recovery epoch _should_ be a subset of $S$
   * So we select the subset of all collectors $s_c$ from $S$. This are the collectors that can participate in the recovery epoch and we can partition them across the different clusters (method `ConstructClusterAssignment`)
   * For a cluster to work, we need a root QC, for the cluster to start its local HotStuff. Hence, each cluster must contain _more_ than 2/3 of collectors (by weight), whose private staking key we know. 
   * Therefore, we split  $s_c$ into two subsets 
      * $s^{(i)}_c$ are the nodes, where we know the private staking key: 
      * $s^{(p)}_c$ are the nodes, where we do _not_ know the private staking key 

      We can proceed as follows: 
       1. We load the old bootstrapping data for our internal nodes info from disk (step `ReadInternalNodeInfos`). Then we iterate over all nodes $n \in s_c$. If collector node $n$ is listed in the old bootstrapping data, we have a private staking key for it and it goes into $s^{(i)}_c$. Otherwise, the collector is added to $s^{(p)}_c$ (no private key). 
       2. we can then call (similar to [line 124](https://github.com/onflow/flow-go/blob/7d09f47d73fbf7930a38d2e591d783209c7f92a2/cmd/util/cmd/epochs/cmd/recover.go#L124)) `common.ConstructClusterAssignment(` `log`, $s^{(p)}_c$, $s^{(i)}_c$, `flagCollectionClusters` `)`

To the best of my understanding, this is the most general algorithm and works for the broadest possible range of situations. In particular:
* If is possible for us to add or remove internal collectors _after_ the original spork. 

   In comparison, removing one of our collectors would yield an error in Jordan's suggestion: 
   > Exit with an error if we observe a node in InternalNodeInfos that doesn't exist in the snapshot

Reply:
@AlexHentschel thanks for the in depth algorithm clarity

https://github.com/onflow/flow-go/pull/5576/commits/2b9c152bcf90a152060beba47b28646c5c1d2ee3